experimental/priority_sampler.go

```diff
@@ -0,0 +1,151 @@
+	swapped := false
+	for !swapped {
+		value := atomic.LoadUint64(field)
+		swapped = atomic.CompareAndSwapUint64(field, value, value|mask)
+	}
+}
+
+// PrioritySampler contains a list of samplers that it interrogates in order.
+// Sampling methods return as soon as one of the samplers returns sample=true.
+// The retryable state for each underlying sampler is stored in the extended context
+// and once retryable=false is returned by one of the delegates it will never be
+// called again.
+type PrioritySampler struct {
+	jaeger.SamplerV2Base
+
+	delegates []jaeger.SamplerV2
+}
+
+type extendedStateKey string
+
+// TODO this may need to be unique per instance of the sampler.
+const prioritySamplerExtendedStateKey = extendedStateKey("priority-sampler")
+
+// NewPrioritySampler creates a new PrioritySampler with given delegates.
+func NewPrioritySampler(delegates ...jaeger.SamplerV2) *PrioritySampler {
+	return &PrioritySampler{delegates: delegates}
+}
+
+// OnCreateSpan implements SamplerV2 API.
+func (s *PrioritySampler) OnCreateSpan(span *jaeger.Span) jaeger.SamplingDecision {
+	return s.trySampling(
+		span,
+		func(sampler jaeger.SamplerV2) jaeger.SamplingDecision {
+			return sampler.OnCreateSpan(span)
+		},
+	)
+}
+
+// OnSetOperationName implements SamplerV2 API.
+func (s *PrioritySampler) OnSetOperationName(span *jaeger.Span, operationName string) jaeger.SamplingDecision {
+	return s.trySampling(
+		span,
+		func(sampler jaeger.SamplerV2) jaeger.SamplingDecision {
+			return sampler.OnSetOperationName(span, operationName)
+		},
+	)
+}
+
+// OnSetTag implements SamplerV2 API.
+func (s *PrioritySampler) OnSetTag(span *jaeger.Span, key string, value interface{}) jaeger.SamplingDecision {
+	return s.trySampling(
+		span,
+		func(sampler jaeger.SamplerV2) jaeger.SamplingDecision {
+			return sampler.OnSetTag(span, key, value)
+		},
+	)
+}
+
+// OnFinishSpan implements SamplerV2 API.
+func (s *PrioritySampler) OnFinishSpan(span *jaeger.Span) jaeger.SamplingDecision {
+	return s.trySampling(
+		span,
+		func(sampler jaeger.SamplerV2) jaeger.SamplingDecision {
+			return sampler.OnFinishSpan(span)
+		},
+	)
+}
+
+// Close calls Close on all delegate samplers.
+func (s *PrioritySampler) Close() {
+	for _, s := range s.delegates {
+		s.Close()
+	}
+}
+
+func (s *PrioritySampler) getState(span *jaeger.Span) *prioritySamplerState {
+	ctx := span.Context().(jaeger.SpanContext)
+	return ctx.ExtendedSamplingState(
+		prioritySamplerExtendedStateKey,
+		func() interface{} {
+			return newPrioritySamplerState(len(s.delegates))
+		},
+	).(*prioritySamplerState)
+}
+
+func (s *PrioritySampler) trySampling(
+	span *jaeger.Span,
+	sampleFn func(v2 jaeger.SamplerV2) jaeger.SamplingDecision,
+) jaeger.SamplingDecision {
+	state := s.getState(span)
+	retryable := false
+	for i := 0; i < len(s.delegates); i++ {
+		if state.isFired(i) {
+			continue
+		}
+		decision := sampleFn(s.delegates[i])
+		if !decision.Retryable {
+			state.setFired(i)
```

Comment:
is `Fired` a descriptive enough name? Would it be better to have something that implied that this sampler is done processing?

Reply:
changed